src/backend/smallbuddyrange.h

```diff
@@ -0,0 +1,226 @@
+#pragma once
+
+#include "../ds/address.h"
+#include "../pal/pal.h"
+#include "range_helpers.h"
+
+namespace snmalloc
+{
+  /**
+   * struct for representing the redblack nodes
+   * directly inside the meta data.
+   */
+  struct FreeChunk
+  {
+    capptr::Chunk<FreeChunk> left;
+    capptr::Chunk<FreeChunk> right;
+  };
+
+  /**
+   * Class for using the allocations own space to store in the RBTree.
+   */
+  class BuddyInplaceRep
+  {
+  public:
+    using Holder = capptr::Chunk<FreeChunk>;
+    using Contents = capptr::Chunk<FreeChunk>;
+
+    static constexpr Contents null = nullptr;
+
+    static constexpr address_t MASK = 1;
+    static void set(Holder* ptr, Contents r)
+    {
+      SNMALLOC_ASSERT((address_cast(r) & MASK) == 0);
+      if (r == nullptr)
+        *ptr = capptr::Chunk<FreeChunk>(
+          reinterpret_cast<FreeChunk*>(address_cast(*ptr) & MASK));
```

Comment:
The CHERI compiler correctly flags this as the construction of a pointer from a provenance-free integer.
```suggestion
          reinterpret_cast<FreeChunk*>((*ptr).unsafe_uintptr() & MASK));
```